libraries/default.rb

```diff
@@ -40,6 +40,44 @@ def self.error_message(error_code)
     end
   end
 
+  class ZookeeperConfig < Hash
+    IMMUTABLE_FIELDS = %w(dynamicConfigFile).freeze
+
+    def self.from_h(input)
+      ZookeeperConfig.new.merge(input)
+    end
+
+    def self.from_text(input)
+      out = ZookeeperConfig.new
+      input.split("\n").each do |line|
+        c = line.split('=')
+        out.merge!({ c[0] => c[1] })
+      end
+      out
+    end
+
+    def apply!(source)
+      update = source.reject { |k, _| IMMUTABLE_FIELDS.include?(k) }
+      transform_values!.with_index do |v, i|
+        k = keys[i]
+        if IMMUTABLE_FIELDS.include?(k)
+          v
+        elsif !update.key?(k)
+          nil
+        else
+          update.delete(k)
+        end
+      end
+      compact!
+      merge!(update)
+      self
+    end
+
+    def to_s
```

Comment:
OPT: maybe we could simplify all its logic by saying the method to_s of this is:
- a sort_by of fields using the position they appear in the existing config file
- replacing the value of the immutable fields

in pseudo code:

```
def to_s
  self.sort_by { |k,v| input_file.lines.find_index { |l| l =~ /^#{k}=/ } }.map { "#{k}=#{v}" }
end
```

it would allow to drop most of the apply! method.

What do you think?

Reply:
I don't like the idea to mix logic and representation in #to_s 
But I can maybe simplify #apply! logic anyway

Reply:
I made a proposal, wdyt?

Reply:
not sure to understand yet how the order is kept

Reply:
When I switched from Array to Hash storage in that class, I had a look and apparently Hash is an ordered structure. Keys are ordered in the order of insertion.